app.py

```diff
@@ -1,150 +1,269 @@
+    assert waii_tools_descriptions == langchain_tools_descriptions
+
+    return langchain_tools
+
+
+def _get_langchain_tools(llm: BaseLanguageModel) -> List[BaseTool]:
+    conn_string = (f'snowflake://{st.secrets["user"]}:{st.secrets["password"]}'
+                   f'@{st.secrets["account_identifier"]}/{st.secrets["database_name"]}/{st.secrets["schema_name"]}'
+                   f'?role={st.secrets["user"]}&warehouse={st.secrets["warehouse_name"]}')
+    return SQLDatabaseToolkit(db=SQLDatabase.from_uri(conn_string), llm=llm).get_tools()
+
 
+def _create_agent(model: _Model, toolkit: _Toolkit) -> AgentExecutor:
+    llm = ChatOpenAI(model=model.value, temperature=0, streaming=True)
+    if toolkit == _Toolkit.WAII:
+        tools = _get_waii_tools()
+        system_prompt = SystemMessagePromptTemplate(prompt=kai_gen_sql)
+    elif toolkit == _Toolkit.LANGCHAIN:
+        tools = _get_langchain_tools(llm)
+        system_prompt = SystemMessagePromptTemplate(prompt=custom_gen_sql)
+    else:
+        raise ValueError(f'Unknown toolkit: {toolkit}')
 
-# Function to handle user feedback
-def handle_feedback(feedback_type):
-    feedback_counts[feedback_type] += 1
+    prompt: ChatPromptTemplate = ChatPromptTemplate.from_messages([
+        system_prompt,
+        MessagesPlaceholder(variable_name='chat_history'),
+        HumanMessagePromptTemplate(prompt=PromptTemplate(input_variables=['input'], template='{input}')),
+        MessagesPlaceholder(variable_name='agent_scratchpad')
+    ])
+    chat_memory = StreamlitChatMessageHistory(key=_ST_CHAT_HiSTORY_KEY)
+    chat_memory.clear()
+
+    return AgentExecutor(
+        agent=create_openai_tools_agent(llm, tools, prompt=prompt),
+        tools=tools,
+        verbose=True,
+        memory=ConversationBufferMemory(memory_key='chat_history', chat_memory=chat_memory, return_messages=True)
+    )
 
 
-chat_display = ChatDisplay(agent_executor)
-chat_display.display_chat()
+def _call_agent(agent: AgentExecutor, user_input: str) -> Mapping[str, Any]:
+    return agent.invoke(
+        input={'input': user_input},
+        config={'callbacks': [
+            StreamlitCallbackHandler(st.container(), expand_new_thoughts=True),
+            LLMonitorCallbackHandler(app_id=st.secrets.LUNARY_APP_ID),
+            FileCallbackHandler('chat_log.txt'),
+        ]}
+    )
 
-view_messages = st.sidebar.expander('View the message contents in session state')
+class _AI_config(BaseModel):
+    ai_tr_name: str = Field(description="name of the transformation (with spaces between words)")
+    ai_tr_description: str = Field(description="description of the transformation")
+    ai_output_table: str = Field(description="name of the output table (no spaces)")
 
-# get the output of the last message from the agent 
-if len(msgs_history.messages) > 1:
-    last_output_message = msgs_history.messages[-1].content
+def generate_config_details(chat_history):
+    model = ChatOpenAI(model='gpt-4-1106-preview', temperature=0)
```

Comment:
I think that you should use the same model that the user selected for the chatbot. It is stored in `st.session_state.model`, but it might be better to pass it to this function as a parameter from the `app()` function.

Reply:
I've fixed that in 11f5fc0.